tests/test_debug/test_snapshot_manager.py

```diff
@@ -0,0 +1,151 @@
+from __future__ import annotations
+
+import os.path
+import shutil
+import sys
+
+import pytest
+from pytest_snapshot.plugin import Snapshot
+
+from python_ta.debug import SnapshotManager
+
+SNAPSHOT_DIR = os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), "snapshot_manager_testing_snapshots"
+)
+TEST_RESULTS_DIR = "/tmp/test_results"
```

Comment:
pytest provides a fixture for creating a temporary directory, use that instead of defining your own here

Reply:
I defined a directory here because the output directory is captured in the snapshots, so we want the directories to be consistent across each test run. 

I looked into using [`tmp_path` and `tmp_path_factory`](https://docs.pytest.org/en/stable/how-to/tmp_path.html#how-to-use-temporary-directories-and-files-in-tests) but I can't seem to get the same base output directory for each run (an example directory I got is `/private/var/folders/rt/x1pynv8j1fb1hkbhs6y9v62m0000gn/T/pytest-of-leoliu/pytest-0`). The only way I found to ensure the base directory is the same is to set the `--basetemp`. Should I rely on this flag to ensure the directories a re consistent?

Reply:
@leowrites that's a great question. This by itself suggests a limitation with the tests, because (of course) you mean to only test the part inside the `with` block, and not this `output_directory` parameter.

I think you should be able to change your function structure by nesting the "function input" definition within the test case itself, making the `output_directory` variable non-local to the function. Something like:

```python
def test_func_one_line(tmp_dir)
    output_directory = ... # tmp_dir's path
    
    def func_one_line():
        with SnapshotTracer(...):
            num = 123

    func_one_line()

    assert_output_files_match(..., ...)
```

This loses the nice use of `pytest.parameterize` that you have, but I think overall the code complexity stays almost the same because you move the function input definitions into the test case code.

Reply:
@david-yz-liu By nesting the functions, the `output_directory` is still captured in `f_locals` inside `func_one_line`. I think this is because `f_locals` also includes free variables in the parent scope when they are referenced. One solution I found is using a global variable so that the variable won't show up in `f_locals`, but it feels unconventional and could break if we were to run the tests concurrently

Reply:
@leowrites wow, yeah okay. I'm not sure how I feel about free variables appearing the stack frame, but I get why it's happening.

All that said, it's not good to use an absolute path (can't assume specific permissions) and it's almost certainly going to look different on Windows.

Why don't we take a pause on this and introduce a different approach. One `snapshot` feature that's on our roadmap is to allow the user to _exclude_ certain variables (again, as an iterable string/regex). It's a different kind of filtering, I suppose, now at the individual frame level. If you implement that, you'll be able to use it here to exclude `output_directory` from the memory model diagram

Reply:
@david-yz-liu This approach worked and I was able to remove the hard coded test directory! 

One thing I noticed while testing is when there are no variables in any frames, `snapshot` calls the memory-viz CLI with an empty JSON `[]`. The memory-viz CLI then errors `Error: Reduce of empty array with no initial value` and the program crashes. 

I added [two checks](https://github.com/pyta-uoft/pyta/pull/1086/files#diff-c6c9e24243fc4ffd8248c423e1c47db81d2340a6d8733bc9cf627054b6169bcdR87-R95) in `snapshot` so that we don't invoke the CLI when the JSON is empty. What do you think about this?

Reply:
@leowrites glad to hear this approach worked.

What you're describing is a bug (or perhaps multiple bugs) with MemoryViz itself. I don't want you to add checks in the `snapshot` code here (it should be fine if MemoryViz receives no objects/stack frames). If you have any test cases that are causing this error, you can mark them pending for now until you fix the MemoryViz bug (and I make a new release).